src/block/client/posix/block_client.c

```diff
@@ -1,6 +1,144 @@
 
+#include <stdio.h>
 #include <stdlib.h>
 #include <block_client.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <string.h>
+#include <errno.h>
+#include <signal.h>
+
+size_t ring_alloc(ring_t *r, uint64_t buffer_size)
+{
+    r->size = buffer_size ? buffer_size / sizeof(struct ctrl) : _SC_AIO_LISTIO_MAX - 1;
+    r->buffer = malloc(r->size * sizeof(struct ctrl *));
+    if(r->buffer){
+        for(size_t i = 0; i < r->size; i++){
+            r->buffer[i] = malloc(sizeof(struct ctrl));
+            if(!r->buffer[i]){
```

Comment:
Shouldn't we be deallocating r->buffer[0..i-1] on error to avoid leaks?

Reply:
If `ring_alloc` fails and returns `0` `block_client_finalize` will be called which traverses through the whole structure and deallocates it. That's why `r->buffer` isn't freed either if the first `malloc` fails.